### ISSUE TITLE
Kafka Engine improvements RFC

### ISSUE THREAD

**Issue by filimonov** [ClickHouse/ClickHouse#65297]
Kafka Engine in ClickHouse is the most convenient way for ClickHouse-Kafka integration that does not require any extra components. It has been contributed to ClickHouse by Cloudflare back in 2018. Since then it has been battle tested and used by thousands of ClickHouse installations. Yet, it is not ideal, and can certainly be improved. Below is a list of improvements that we propose to do and are going to work on.

## Better integration with Confluent infrastructure

* Create ProtobufConfluent similar to AvroConfluent - see https://github.com/ClickHouse/ClickHouse/issues/67675#issuecomment-2392302653

## Error handling

Error handling in the Kafka engine is inconvenient and requires a lot of care. One malformed message may block consumers. Following improvements can make it better:

* Add exponential backoff for retries, expose next retry in kafka_consumers table. One failing consumer can currently hurt the whole cluster, because it will be leaving/entering consumer group all the time, which is ‘stop the world’ operation. Example bug report: [https://github.com/ClickHouse/ClickHouse/issues/33433](https://github.com/ClickHouse/ClickHouse/issues/33433) 
* Add two more modes into `kafka_handle_error_mode`:
    1. ‘Ignore’ – ignore all errors completely, just increase error counters
    2. ‘dead_letter_queue‘. This mode would redirect the errors to a dedicated system table `system.dead_letter_queue` (similar to `_log` tables there). Proposed structure: 
   ```
   event_time,
   database,
   table,
   topic,
   partition,
   offset,
   raw_message,
   error
   ```
* Add exponential slow-down of retries if kafka can’t get assignment: [https://github.com/ClickHouse/ClickHouse/issues/42803](https://github.com/ClickHouse/ClickHouse/issues/42803) 

## Clean startup/shutdown

The initialization sequence of different table types in ClickHouse is not deterministic. Therefore it is possible that the Kafka engine starts consuming before the destination table is ready. Same is true for stopping the server. That leads to unnecessary errors in the log and potential consistency issues. Example bug report: [https://github.com/ClickHouse/ClickHouse/issues/48926](https://github.com/ClickHouse/ClickHouse/issues/48926)

* During startup: postpone the start of consumption until all tables and MVs are initialized
* During shutdown: stop the consumption before any table is shut down 

## Usability

Following improvements can improve usability of Kafka engine in general:

* Round-robin consumption of several partitions. Currently in case of big lag ClickHouse will consume partitions one after another, i.e. the first partition till the end, then the second and so on. Because of that the maximum lag stays high all the time, and the lag in time (not in messages) starts decreasing only after the last partition starts to be consumed. See [https://github.com/mfontanini/cppkafka/blob/9c5ea0e332486961e612deacc6e3f0c1874c688d/tests/roundrobin_poll_test.cpp#L72](https://github.com/mfontanini/cppkafka/blob/9c5ea0e332486961e612deacc6e3f0c1874c688d/tests/roundrobin_poll_test.cpp#L72)
* Add table function kafka() for schema inference: `DESCRIBE kafka(...)`. Alternatively: research other options for ‘smart’ / ‘automatic’ schema inference & evolution.
* Allow to manage consumer group offsets in SQL (e.g. to skip something, retry something, rewind back etc.). It can be implemented via SYSTEM commands or ALTER TABLE MODIFY SETTING. See some context there: [https://gist.github.com/filimonov/1646259d18b911d7a1e8745d6411c0cc](https://gist.github.com/filimonov/1646259d18b911d7a1e8745d6411c0cc) 

    [https://github.com/confluentinc/librdkafka/blob/master/examples/alter_consumer_group_offsets.c](https://github.com/confluentinc/librdkafka/blob/master/examples/alter_consumer_group_offsets.c)

## Housekeeping

There are a number of small things that can be done in order to make implementation cleaner.

* Make message_broker_pool expanded automatically on demand (this is also true for NATS & RabbitMQ). Current implementation is confusing and error prone. It reuses the schedule thread pool to simplify rescheduling of the tests (static number of threads is strange here).
* Remove some old / confusing settings like commit_every_batch, improve documentation that confuse users. Add documentation and create examples with named_collections for librdkafka settings (see [https://github.com/ClickHouse/ClickHouse/pull/59710](https://github.com/ClickHouse/ClickHouse/pull/59710))
* Upgrade to librdkafka 2.4.1 (maybe it has the bug with shutting down the consumer fixed).  \
A potential fix: [https://github.com/confluentinc/librdkafka/pull/4724](https://github.com/confluentinc/librdkafka/pull/4724) \
One more report (of many): [https://github.com/confluentinc/librdkafka/issues/4674](https://github.com/confluentinc/librdkafka/issues/4674) \
Motivation [https://github.com/ClickHouse/ClickHouse/issues/41216](https://github.com/ClickHouse/ClickHouse/issues/41216) 

Fix minor bugs:

* [https://github.com/ClickHouse/ClickHouse/issues/39333](https://github.com/ClickHouse/ClickHouse/issues/39333) one of the metrics can become negative
* [https://github.com/ClickHouse/ClickHouse/issues/41335](https://github.com/ClickHouse/ClickHouse/issues/41335) (better logging).
* [https://github.com/ClickHouse/ClickHouse/issues/55200](https://github.com/ClickHouse/ClickHouse/issues/55200) Kerberos - update ticket w/o restart (seems trivial, just allow sasl_kerberos_min_time_before_relogin)

## Exactly-once message delivery (EOS)

The current implementation of Kafka Engine guarantees at-least-once semantics. That means a message is guaranteed to be inserted into ClickHouse, but in rare cases it can be inserted twice. 

It happens due to limitations of the current implementation. When Kafka engine consumes messages there are two commits performed into two different distributed systems: first commit is data flush into ClickHouse, second – commit offset into Kafka. That makes it impossible to guarantee exactly once semantics: if Kafka becomes unavailable while we are flushing data into ClickHouse we will not be able to commit and can get duplicates (it’s a variation of ‘[Two Generals' Problem](https://en.wikipedia.org/wiki/Two_Generals%27_Problem)). In order to guarantee exactly once message delivery, there should be a single commit, and offsets should be committed together with data.

Following incremental improvements are proposed in order to reduce or completely eliminate risk of duplicates:

* consume every partition separately (round robin or by different threads)
* Have EOS logic enabled in librdkafka (isolation_level=read_committed), have sticky or rack-aware consumers enabled to avoid jumping the partitions between shards
* enable the possibility to control the offsets of the subscribed topic manually (it’s just ‘assign’ command: [https://github.com/ClickHouse/ClickHouse/blob/dc2d5e0ec65e34a14d37bd030f32117a11d08866/src/Storages/Kafka/KafkaConsumer.cpp#L419](https://github.com/ClickHouse/ClickHouse/blob/dc2d5e0ec65e34a14d37bd030f32117a11d08866/src/Storages/Kafka/KafkaConsumer.cpp#L419) ) and possibility to extract the block of messages of the desired size. Alternatively we can try to use ‘old-style’ API to consume from Kafka. 
* Store the coordinates (i.e. range) of the message block from the partition in the (zoo)keeper before writing it to the MV. 
* Commit the successful write to (zoo)keeper, or retry the same block on failure.
* deduplicate_blocks_in_dependent_materialized_views + generate block id? 

## Research

There are some areas that require additional research before committing any work.

* Benchmarks consumer performance for different formats and investigate if it can be improved. For example, there are several reports that kafka_format = 'ProtobufSingle' is too slow.  [https://github.com/filimonov/ch-kafka-consume-perftest](https://github.com/filimonov/ch-kafka-consume-perftest) 
    * The parsing performance is much better when there are a lot of data rows in a single Kafka message (for example 10x for Protobuf). Probably it’s caused by slow parser re-initalization after every message. Most probably it can be improved (most probably async inserts will also benefit from that, as they share parts of the code here). \
Some work has been done in this PR but postponed: [https://github.com/ClickHouse/ClickHouse/pull/59482](https://github.com/ClickHouse/ClickHouse/pull/59482)
* Repro is needed:
    * Local: No offset stored: [https://github.com/ClickHouse/ClickHouse/issues/12295](https://github.com/ClickHouse/ClickHouse/issues/12295)
    * [https://github.com/ClickHouse/ClickHouse/issues/56589](https://github.com/ClickHouse/ClickHouse/issues/56589) some regression related to pipeline
    * [https://github.com/ClickHouse/ClickHouse/pull/63039/files](https://github.com/ClickHouse/ClickHouse/pull/63039/files) Librdkafka code have some data-race in the stat calculation (suppression, maybe can be fixed?) 
    * Some issue between Protobuf & Kafka [https://github.com/ClickHouse/ClickHouse/issues/41971](https://github.com/ClickHouse/ClickHouse/issues/41971)
    * [https://github.com/ClickHouse/ClickHouse/issues/37877](https://github.com/ClickHouse/ClickHouse/issues/37877) research / document sticky consumers / rack-aware consumers (so the partition will never be moved between shards).
* Possibility to backup the offsets? 
* Kafka 4 support https://github.com/ClickHouse/ClickHouse/issues/63697
* Kafka engine as a producer: 
    * In general the code is immature, and may have a lot of different issues. Can work incorrectly, slow and  strange with some formats. Research and excessive testing is needed. Currently the Kafka engine as a producer can be used only for simple cases like signaling some anomalies (i.e. sending a few messages sometimes, not a big stream).
    * [https://github.com/ClickHouse/ClickHouse/issues/41204](https://github.com/ClickHouse/ClickHouse/issues/41204)

* Support for IAM auth for MSK (via SASL_OAUTHBEARER, see https://github.com/vectordotdev/vector/issues/1516#issuecomment-1826482110 + https://github.com/ClickHouse/ClickHouse/pull/43259)

References:

* https://gist.github.com/filimonov/45c45974cf568c43f62366001c2e3857
* https://gist.github.com/filimonov/c1dce9b7e560f20cb4d1de5b930ca93f
* https://gist.github.com/filimonov/791ad14908830784ccdeee113074f644

**Comment by antaljanosbenjamin** [comment]
I am already working on exactly once semantics [in this PR](https://github.com/ClickHouse/ClickHouse/pull/57625). My goal is to create a new experimental StorageKafka engine that stores the offsets in keeper and consumes only from a single partition at a time (on a single thread, so multiple threads will consume from multiple partitions simultaneously). I achieved this by manipulating the redirection of separate topic queues of rdkafka consumers. I am close to finishing it as most of the tests already working, but there still a few failing.

For the rest of the issue I think also make sense, especially upgrading librdkafka.

**Comment by pet74alex** [comment]
> I am already working on exactly once semantics [in this PR](https://github.com/ClickHouse/ClickHouse/pull/57625).
@antaljanosbenjamin Could you please make possible to use your Kafka engine improvements in the keeperless configurations?

**Comment by antaljanosbenjamin** [comment]
> Could you please make possible to use your Kafka engine improvements in the keeperless configurations?

First of all, the new engine is still experimental. If you opt for testing it out, that makes me happy, but be careful and please don't put the new engine into production use where issues can cause headaches for you. There might be some big issue lurking under the hood.

Second, as the goal of the new table engine to store offsets in Keeper, it is impossible to make it work without Keeper.

**Comment by ilejn** [comment]
Hello @filimonov , could you clarify why _stream_ and _dead_letter_queue_ should coexist.
Is not it true that dead_letter_queue table can be maintained by _stream_ and a bunch of materialized views?
If we treat _dead_letter_queue_ like a "syntax sugar", is it fair to give user a choice between these two  modes disabling their combination?

**Comment by filimonov** [comment]
> Hello @filimonov , could you clarify why _stream_ and _dead_letter_queue_ should coexist. Is not it true that dead_letter_queue table can be maintained by _stream_ and a bunch of materialized views? If we treat _dead_letter_queue_ like a "syntax sugar", is it fair to give user a choice between these two modes disabling their combination?

It's similar, but using stream is quite complex and requite the end user to know how to setup that properly. dead_letter_queue - should be much simpler for the user, is kinda syntax sugar. Nobody want to complicate their pipelines with some error processing materialzied views. 

Also it can be done a little bit more elegant in the code (just another system log table, no need to populate _error and put it to a stream of normal messages).

**Comment by antaljanosbenjamin** [comment]
Maybe consider `kafka_dead_letter_queue` as a table name. In case of the future we want to use a dead letter queue for another streaming engine which is not compatible with the schema of the proposed table.

**Comment by ilejn** [comment]
@filimonov , @alex-zaitsev  ,  please reconfirm that we do need `ignore` for `kafka_handle_error_mode`, despite the fact we already have `input_format_allow_errors_ratio` .

**Comment by filimonov** [comment]
> input_format_allow_errors_ratio

It's completely different, and supported only by certain formats (based on IRowInputFormat).

**Comment by antaljanosbenjamin** [comment]
Pasting the same information here as [it was in my PR](https://github.com/ClickHouse/ClickHouse/pull/57625#issuecomment-2268964711).

So the PR #57625 was merged, so we have the first, experimental version of Kafka engine in ClickHouse that stores the offsets in ZooKeeper. In order to make it on par with the current Kafka engine and production ready, we have to do _at least_ the following things:
1. Communicating the reason why restarting the table didn't work, for more info see https://github.com/ClickHouse/ClickHouse/pull/57625#discussion_r1668580042 .
2. Make sure table recreation with same ZK path doesn't mess up the table, so saving the metadata of the table to ZK to make sure we deal with the exact same table (even if UUID matches, the topics might be different).
3. Support direct read.
4. Support `system.kafka_consumers` table.

**Comment by spff** [comment]
Hope the improvement also helps the unstable / timeout issues 🙏  https://github.com/ClickHouse/ClickHouse/issues/66415

**Comment by antaljanosbenjamin** [comment]
I managed to update librdkafka (will push the changes tomorrow) with some minor changes in our CMake, but the mentioned issue (https://github.com/confluentinc/librdkafka/pull/4724) is a must have for the new Kafka engine, `test_kafka_consumer_failover` hangs without that. If the tables are dropped before the topic is removed or with the PR the test succeeds.

**Comment by ilejn** [comment]
> I managed to update librdkafka (will push the changes tomorrow) 

Hello @antaljanosbenjamin ,  did you face https://github.com/confluentinc/librdkafka/issues/4782  ?

**Comment by antaljanosbenjamin** [comment]
I haven't run TSAN builds or a lot of tests, just a few ones locally. Those were fine, but it might pop up in CI.

**Comment by ilejn** [comment]
I kind of got stuck with https://github.com/ClickHouse/ClickHouse/pull/65314

**Comment by antaljanosbenjamin** [comment]
So my last message was from my phone, that is why it was short. I managed to build-compile everything locally. I tried different versions of librdkafka, but all of the 2.x.x versions seem to have the bug https://github.com/confluentinc/librdkafka/pull/4724 , which makes them a no-go. The latest 1.9.2 version looked promising, I will try to upgrade our version to that. If that works out of the box, then let's go with that first and then I will try to make the 2.x.x version work. Depending on the quality of 1.9.2, it might be high or low priority.

**Comment by ilejn** [comment]
@antaljanosbenjamin ,  to me it is acceptable to create custom version (it was actually what I did https://github.com/ilejn/librdkafka/commits/branch_2.4.0) . I've fixed a couple of issues found by sanitizers, but not https://github.com/confluentinc/librdkafka/pull/4724
2.x.x is not a must have though, agree.

**Comment by antaljanosbenjamin** [comment]
So I continued my journey. The version 1.8.0 contains a bug that makes the old kafka engine hanging during commit if the topic is removed or the broker is unavailable. Basically https://github.com/confluentinc/librdkafka/issues/3740, but with sync commit call. For version 1.7.0 the default value of `session.timeout.ms` is changed from 10s to 45s. I had to change that to make rebalance/failover related tests pass. I will make a PR for upgrade to 1.7.0, but I don't see the way how to proceed further because of the issues with librdkafka.

**Comment by antaljanosbenjamin** [comment]
Even [upgrading to 1.7.0](https://github.com/ClickHouse/ClickHouse/pull/69740) is problematic: https://github.com/confluentinc/librdkafka/pull/3340 breaks our kerberized kafka tests.

**Comment by jovezhong** [comment]
> Support for IAM auth for MSK

While Java, Python, Go, and other languages got official or community SDKs for the IAM role support, I don't think there is a ready-to-use C++ lib for this. So we built our own and made them available at 
- https://github.com/timeplus-io/proton/blob/develop/src/IO/Kafka/AwsMskIamSigner.h
- https://github.com/timeplus-io/proton/blob/develop/src/IO/Kafka/AwsMskIamSigner.cpp

With that, `AWS_MSK_IAM` is available for `sasl_mechanism` setting and you can read/write data in AWS MSK via
```sql
CREATE EXTERNAL STREAM msk_stream(column_defs)
SETTINGS
    type='kafka', topic='topic2',
    brokers='prefix.kafka.us-west-2.amazonaws.com:9098',
    security_protocol='SASL_SSL',
    sasl_mechanism='AWS_MSK_IAM';
SELECT * FROM msk_stream;
```

While it's possible to contribute the code to AWS SDK for C++, maybe it'll be easier to put them in ClickHouse codebase, since our code already uses some S3/Base64 wrappers from ClickHouse.

**Comment by filimonov** [comment]
After https://github.com/ClickHouse/ClickHouse/pull/76621 we can take a create some OAUTH  look on AWS IAM 
https://aws.amazon.com/blogs/big-data/amazon-msk-iam-authentication-now-supports-all-programming-languages/

And maybe on google managed Kafka
https://cloud.google.com/managed-service-for-apache-kafka/docs/authentication-kafka

**Comment by lesandie** [comment]
https://github.com/ClickHouse/ClickHouse/pull/76621

**Comment by filimonov** [comment]
Update kafka images in the tests (maybe move to > 4.0 & avoid starting zookeeper, as it fail sometimes):
https://s3.amazonaws.com/clickhouse-test-reports/json.html?REF=master&sha=f4ade7f75a7ef44366687b65a2103df1608228a1&name_0=MasterCI&name_1=Integration%20tests%20%28tsan%2C%202%2F6%29

**Comment by alexey-milovidov** [comment]
Outdated.
